File: fidgetswipe-ios/fidgetswipe-ios/Action.swift
//
//  Action.swift
//  fidgetswipe-ios
//
//  Created by Bradley Mackey on 07/07/2017.
//  Copyright © 2017 Bradley Mackey. All rights reserved.
//

import Foundation

/// A simple enum to represent the type of actions that the game accepts.
public enum Action: Int, CustomStringConvertible {
    
    case tap        = 0
    case swipeUp    = 1
    case swipeDown  = 2
    case swipeRight = 3
    case swipeLeft  = 4
    case shake      = 5
    case upsideDown = 6
    case volumeUp   = 7
    case volumeDown = 8
    case faceUp     = 9
    case faceDown   = 10
    
    case timeRanOut = 999 // special action, only called when the time has run out.
    
    /// The total number of actions in the enum.
    public static let totalActions:UInt32 = 11
    
    public static func random() -> Action {
        let randomNumber = arc4random_uniform(Action.totalActions)
        if let action = Action(rawValue: Int(randomNumber)) {
            return action
        } else {
            fatalError("rawValue (\(randomNumber)) does not equate to a vaild 'Action'.")
        }
    }
    
    public var description:String {
        switch self {
        case .tap:
            return "TAP"
        case .swipeUp, .swipeDown, .swipeLeft, .swipeRight:
            return "SWIPE"
        case .shake:
            return "SHAKE"
        case .upsideDown:
            return "ROTATE"
        case .volumeUp:
            return "VOLUME UP"
        case .volumeDown:
            return "VOLUME DOWN"
        case .faceUp:
            return "FACE UP"
        case .faceDown:
            return "FACE DOWN"
        case .timeRanOut:
            fatalError("Time ran out is not an action that should be displayed")
        }
    }
    
    public var isMotionChallenge:Bool {
        switch self {
        case .upsideDown, .faceUp, .faceDown, .shake:
            return true
        case .tap, .swipeRight, .swipeLeft, .swipeDown, .swipeUp, .volumeUp, .volumeDown, .timeRanOut:
            return false
        }
    }
}

File: FidgetPhone-android/app/src/main/java/com/bradleymackey/fidgetphone/model/Action.java
package com.bradleymackey.fidgetphone.model;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public enum Action {

    TAP,
    SWIPE_UP,
    SWIPE_DOWN,
    SWIPE_RIGHT,
    SWIPE_LEFT,
    SHAKE,
    UPSIDE_DOWN,
    VOLUME_UP,
    VOLUME_DOWN,
    FACE_UP,
    FACE_DOWN,
    TIME_RAN_OUT; // the special action

    private static final List<Action> VALUES =
            Collections.unmodifiableList(Arrays.asList(values()));
    private static final int SIZE = VALUES.size();
    private static final Random RANDOM = new Random();

    public static Action random()  {
        Action randomAction = TIME_RAN_OUT;
        while (randomAction == TIME_RAN_OUT) {
            randomAction = VALUES.get(RANDOM.nextInt(SIZE)-1);
        }
        return VALUES.get(RANDOM.nextInt(SIZE));
    }

    public String getDescription() {
        switch (this) {
            case TAP:
                return "TAP";
            case SWIPE_UP: case SWIPE_LEFT: case SWIPE_DOWN: case SWIPE_RIGHT:
                return "SWIPE";
            case SHAKE:
                return "SHAKE";
            case UPSIDE_DOWN:
                return "ROTATE";
            case VOLUME_UP:
                return "VOLUME UP";
            case VOLUME_DOWN:
                return "VOLUME DOWN";
            case FACE_UP:
                return "FACE UP";
            case FACE_DOWN:
                return "FACE DOWN";
            default:
                throw new RuntimeException("Invalid! No description for this Action");
        }
    }

    public boolean isMotionChallenge() {
        switch (this) {
            case UPSIDE_DOWN: case FACE_UP: case FACE_DOWN: case SHAKE:
                return true;
            default:
                return false;
        }
    }

}
